cassovary-core/src/main/scala/com/twitter/cassovary/graph/ArrayBasedDynamicDirectedGraph.scala

```diff
@@ -0,0 +1,206 @@
+
+/**
+ * A dynamic directed graph, implemented using an ArrayBuffer of IntArrayList
+ * (from the fastutil library).  If n nodes are used, O(n) objects are created,
+ * independent of the number of edges.
+ *
+ * For efficiency, it's recommended that ids be sequentially numbered, as
+ * an internal array is stored which is longer than the maximum id seen.
+ *
+ * This class is not threadsafe.  If multiple thread access it simultaneously
+ * and at least one mutates it, synchronization should be used.
+ */
+class ArrayBasedDynamicDirectedGraph(override val storedGraphDir: StoredGraphDir)
+    extends DynamicDirectedGraph {
+  // outboundLists(id) contains the outbound neighbors of the given id,
+  // or null if the id is not in this graph.
+  // If we aren't storing outbound neighbors, outboundLists will always remain size 0.
+  private val outboundLists = new ArrayBuffer[IntArrayList]
+  // (See above note on outboundLists)
+  private val inboundLists = new ArrayBuffer[IntArrayList]
+
+  private var nodeCount_ = 0
+
+  def this(dataIterator: Iterator[NodeIdEdgesMaxId],
+            storedGraphDir: StoredGraphDir) {
+    this(storedGraphDir)
+    for (nodeData <- dataIterator) {
+      val id = nodeData.id
+      getOrCreateNode(id)
+      nodeData.edges map getOrCreateNode
+      storedGraphDir match {
+        case OnlyOut => outboundLists(id).addAll(IntArrayList.wrap(nodeData.edges))
+        case OnlyIn => inboundLists(id).addAll(IntArrayList.wrap(nodeData.edges))
+        case BothInOut => nodeData.edges map { addEdgeAllowingDuplicates(id, _) } // Duplicates shouldn't exist, but allow them for efficiency.
+      }
+    }
+  }
+
+  def this(iteratorSeq: Seq[() => Iterator[NodeIdEdgesMaxId]],
+           storedGraphDir: StoredGraphDir) {
+    this((iteratorSeq map { _()} ).iterator.flatten, storedGraphDir)
+  }
+
+  /* Returns an option which is non-empty if outbound list for id  is non-null. */
+  private def outboundListOption(id: Int): Option[IntArrayList] =
+    if (id < 0 || id >= outboundLists.size)
+      None
+  else
+      Option(outboundLists(id)) // If list is null, convert to None
+
+  private def inboundListOption(id: Int): Option[IntArrayList] =
+    if (id < 0 || id >= inboundLists.size)
+      None
+    else
+      Option(inboundLists(id)) // If list is null, convert to None
+
+  /** Wraps outboundList and inboundList in a node structure.  This is an inner class
+    * because mutations will affect other nodes of the graph if both inbound
+    * and outbound neighbor lists are stored.
+    *
+    * For efficiency, we don't store Nodes, but create them on the fly as needed.
+    */
+  class IntListNode(override val id: Int,
+                    val outboundList: Option[IntArrayList],
+                    val inboundList: Option[IntArrayList])
+      extends DynamicNode {
+    override def outboundNodes(): Seq[Int] = outboundList match {
+      case Some(list) => FastUtilUtils.intArrayListToSeq(list)
```

Comment:
Bear in mind that random access (e.g., for a random walk) would first require traversal of the whole intArrayList to convert the list to a Seq rather than an O(1) direct access. What you want to do I think is to define a new wrapper class like this:
class IntArrayListWrapper(l: IntArrayList) extends IndexedSeq[Int] {
def apply(idx: Int): Int = l.getInt(idx)
def length: Int = l.size
}